sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/HttpInstrumentationOptions.java

```diff
@@ -0,0 +1,303 @@
+     * {@link HttpInstrumentationOptions#addAllowedHeaderName(HttpHeaderName)} or {@link HttpInstrumentationOptions#getAllowedHeaderNames()} to add or
+     * remove more headers names to the existing set of allowed header names.
+     * </p>
+     *
+     * @param allowedHeaderNames The list of allowed header names from the user.
+     *
+     * @return The updated HttpLogOptions object.
+     */
+    public HttpInstrumentationOptions<T> setAllowedHeaderNames(final Set<HttpHeaderName> allowedHeaderNames) {
+        this.allowedHeaderNames = allowedHeaderNames == null ? new HashSet<>() : allowedHeaderNames;
+
+        return this;
+    }
+
+    /**
+     * Sets the given allowed header to the default header set that should be logged when they appear on the request or response.
+     * <p>
+     * Note: headers are not recorded on traces.
+     *
+     * @param allowedHeaderName The allowed header name from the user.
+     *
+     * @return The updated HttpLogOptions object.
+     *
+     * @throws NullPointerException If {@code allowedHeaderName} is {@code null}.
+     */
+    public HttpInstrumentationOptions<T> addAllowedHeaderName(final HttpHeaderName allowedHeaderName) {
+        Objects.requireNonNull(allowedHeaderName);
+        this.allowedHeaderNames.add(allowedHeaderName);
+
+        return this;
+    }
+
+    /**
+     * Gets the allowed query parameters.
+     *
+     * @return The list of allowed query parameters.
+     */
+    public Set<String> getAllowedQueryParamNames() {
+        return Collections.unmodifiableSet(allowedQueryParamNames);
+    }
+
+    /**
+     * Sets the given allowed query params to be recorded on logs and traces.
+     *
+     * @param allowedQueryParamNames The list of allowed query params from the user.
+     *
+     * @return The updated {@code allowedQueryParamName} object.
+     */
+    public HttpInstrumentationOptions<T> setAllowedQueryParamNames(final Set<String> allowedQueryParamNames) {
+        this.allowedQueryParamNames = allowedQueryParamNames == null ? new HashSet<>() : allowedQueryParamNames;
+
+        return this;
+    }
+
+    /**
+     * Sets the given allowed query param that can be recorded on logs and traces.
+     *
+     * @param allowedQueryParamName The allowed query param name from the user.
+     *
+     * @return The updated {@link HttpInstrumentationOptions} object.
+     *
+     * @throws NullPointerException If {@code allowedQueryParamName} is {@code null}.
+     */
+    public HttpInstrumentationOptions<T> addAllowedQueryParamName(final String allowedQueryParamName) {
+        this.allowedQueryParamNames.add(allowedQueryParamName);
+        return this;
+    }
+
+    @Override
+    public HttpInstrumentationOptions<T> setTracingEnabled(boolean isTracingEnabled) {
+        super.setTracingEnabled(isTracingEnabled);
+        return this;
+    }
+
+    @Override
+    public HttpInstrumentationOptions<T> setProvider(T provider) {
```

Comment:
From a non-telemetry background `setProvider` sounds off to me. It feels we're talking about something that'll create an instance of OpenTelemetry but further into code it's actually the OpenTelemetry instance itself.

Is `provider` the terminology used in the telemetry space for the object that'll be capturing traces and spans?

Reply:
I changed it to `set|getTelemetryProvider` - I can't come up with a better name, but happy to think more about alternatives. 
The thing we're setting here is essentially a container that carries telemetry configuration and can be used to create spans, metrics, logs, etc. I.e. it's providing specific implementation that's configured in a certain way. E.g. if we supported micrometer metrics, this would be the `MeterRegistry` (and the implementation would not do any tracing)